js/ui/light.js

```diff
@@ -0,0 +1,256 @@
+                throw new Error('light.color must be a valid premultiplied color array.');
+        }
+
+        this.painter.setLight(colorOpts);
+
+        return this;
+    },
+
+    _validateDirection: function(direction) {
+        if (!Array.isArray(direction) || direction.length !== 3 ||
+            !direction.every(function(i) { return !isNaN(+i); }))
+            throw new Error('light.direction must be an array of three numbers');
+
+        if (direction[1] < 0 || direction[1] >= 360) util.warnOnce('Azimuthal angle will be wrapped to [0, 360)');
+        if (direction[2] < 0 || direction[2] >= 180) util.warnOnce('Polar angle will be clamped to [0, 180)');
+    },
+
+    _sphericalToCartesian: function(r, azimuthal, polar) {
+        // We abstract "north"/"up" to be 0° when really this is 90° (π/2):
+        // correct for that here
+        azimuthal += 90;
+
+        // Convert azimuthal and polar angles to radians
+        azimuthal *= Math.PI / 180;
+        polar *= Math.PI / 180;
+
+        // spherical to cartesian (x, y, z)
+        return [
+            r * Math.cos(azimuthal) * Math.sin(polar),
+            r * Math.sin(azimuthal) * Math.sin(polar),
+            r * Math.cos(polar)
+        ];
+    },
+
+    _setLightDirection: function(direction) {
+        this._validateDirection(direction);
+
+        var r = direction[0],
+            a = direction[1],
+            p = direction[2];
+        var cartesianDirection = this._sphericalToCartesian(r, a, p);
+
+        // We persist both spherical and cartesian coordinates: cartesian for
+        // the painter, spherical for the light APIs (since calculating in
+        // reverse is more difficult)
+        this.painter.setLight({
+            direction: {
+                r: r,
+                a: a,
+                p: p,
+                x: cartesianDirection[0],
+                y: cartesianDirection[1],
+                z: cartesianDirection[2]
+            }
+        });
+
+        return this;
+    },
+
+    _setLightIntensity: function(intensity) {
+        var value = +intensity;
+        if (!isNaN(value) && value >= 0 && value <= 1) {
+            this.painter.setLight({
+                intensity: value
+            });
+        } else throw new Error('light.intensity must be a number between 0 and 1.');
+
+        return this;
+    },
+
+    _setLightOptions: function(mapOpts, styleOpts) {
+        // Used in batch setting initial light options.
+        var lightOpts = util.extend(mapOpts, styleOpts || {});
+        if (lightOpts.anchor) this._setLightAnchor(lightOpts.anchor);
+        if (lightOpts.color) this._setLightColor({rawColor: lightOpts.color});
+        if (lightOpts.direction) this._setLightDirection(lightOpts.direction);
+        if (typeof lightOpts.intensity !== 'undefined') this._setLightIntensity(lightOpts.intensity);
+
+        this._update();
+
+        return this;
+    }
+});
```

Comment:
This class could be simplified by taking advantage of the existing `StyleDeclaration` and `StyleTransition` classes.